src/protocol/Inbox.sol

```diff
@@ -24,36 +26,66 @@ contract Inbox {
     /// @param checkpoint the checkpoint that was proven
     event CheckpointProven(uint256 indexed pubIdx, bytes32 indexed checkpoint);
 
+    error AlreadyProven(uint256 index);
+    error NotProven(uint256 index);
+
+    /// @param bufferSize the maximum number of checkpoints simultaneously stored in this contract
     /// @param genesis the checkpoint describing the initial state of the rollup
     /// @param dataFeed the input data source that updates the state of this rollup
     /// @param verifier a contract that can verify the validity of a transition from one checkpoint to another
-    constructor(bytes32 genesis, address dataFeed, address verifier) {
+    constructor(uint256 bufferSize, bytes32 genesis, address dataFeed, address verifier) {
         // set the genesis checkpoint of the rollup - genesis is trusted to be correct
         require(genesis != 0, "genesis checkpoint cannot be 0");
-        checkpoints[0] = genesis;
+        _checkpoints.setup(bufferSize);
+        _checkpoints.setAt(0, genesis).setLastIndex(1);
```

Comment:
should we use `setLastIndex(0)` instead? The last index is index 0 which is the genesis.